lib/Sema/CSSimplify.cpp

```diff
@@ -6186,6 +6218,91 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyConformsToConstraint(
   return SolutionKind::Error;
 }
 
+ConstraintSystem::SolutionKind ConstraintSystem::simplifyTransitivelyConformsTo(
+    Type type, Type protocolTy, ConstraintLocatorBuilder locator,
+    TypeMatchOptions flags) {
+  // Since this is a performance optimization, let's ignore it
+  // in diagnostic mode.
+  if (shouldAttemptFixes())
+    return SolutionKind::Solved;
+
+  auto formUnsolved = [&]() {
+    // If we're supposed to generate constraints, do so.
+    if (flags.contains(TMF_GenerateConstraints)) {
+      auto *conformance =
+          Constraint::create(*this, ConstraintKind::TransitivelyConformsTo,
+                             type, protocolTy, getConstraintLocator(locator));
+
+      addUnsolvedConstraint(conformance);
+      return SolutionKind::Solved;
+    }
+
+    return SolutionKind::Unsolved;
+  };
+
+  auto resolvedTy = getFixedTypeRecursive(type, /*wantRValue=*/true);
+  if (resolvedTy->isTypeVariableOrMember())
+    return formUnsolved();
+
+  // If the composition consists of a class + protocol,
+  // we can't check conformance of the argument because
+  // parameter could pick one of the components.
+  if (resolvedTy.findIf(
+          [](Type type) { return type->is<ProtocolCompositionType>(); }))
+    return SolutionKind::Solved;
+
+  auto *protocol = protocolTy->castTo<ProtocolType>()->getDecl();
+
+  // First, let's check whether the type itself conforms,
+  // if it does - we are done.
+  if (TypeChecker::conformsToProtocol(resolvedTy, protocol, DC))
+    return SolutionKind::Solved;
+
+  // If the type doesn't conform, let's check whether
+  // an Optional or Unsafe{Mutable}Pointer from it would.
+
+  SmallVector<Type, 4> typesToCheck;
+
+  // Optional<T>
+  typesToCheck.push_back(
+      OptionalType::get(resolvedTy->getWithoutSpecifierType()));
+
+  // Unsafe{Mutable}Pointer<T>
+  {
+    auto &ctx = getASTContext();
+
+    auto *ptrDecl = ctx.getUnsafePointerDecl();
+
+    // String -> UnsafePointer<Void>
+    if (resolvedTy->isEqual(ctx.getStringDecl()->getDeclaredInterfaceType())) {
+      typesToCheck.push_back(BoundGenericType::get(ptrDecl, /*parent=*/Type(),
+                                                   {ctx.TheEmptyTupleType}));
+    }
+
+    // Array<T> -> UnsafePointer<T>
+    if (auto elt = isArrayType(resolvedTy)) {
+      typesToCheck.push_back(
+          BoundGenericType::get(ptrDecl, /*parent=*/Type(), {*elt}));
+    }
+
+    // inout argument -> UnsafePointer<T>, UnsafeMutablePointer<T>
+    if (type->is<InOutType>()) {
+      typesToCheck.push_back(
+          BoundGenericType::get(ptrDecl, /*parent=*/Type(), {resolvedTy}));
+      typesToCheck.push_back(BoundGenericType::get(
+          ctx.getUnsafeMutablePointerDecl(), /*parent=*/Type(), {resolvedTy}));
+    }
```

Comment:
I think you'll also want to handle `AnyHashable` here

Reply:
Good catch! I actually have to add a test for that just like I did for `Optional` and `UnsafePointer`

Reply:
I'm going to let full suite run before pushing that though.

Reply:
Actually I wonder whether I do really need to check for `AnyHashable` here e.g.

```
enum E {
  case test(_: String)
}

func overloaded() -> Int { 42 }
func overloaded() -> E { E.test("") }

func test<T: Hashable>(_: T) -> T {
  fatalError()
}

let _: AnyHashable = test(overloaded())
```

The conversion is from something `Hashable` to `AnyHashable` so the argument is going to check for `Hashable`

Reply:
Right, but if it's a different protocol constraint that `AnyHashable` conforms to, but not the `Hashable` type, then I believe that might be problematic.

Reply:
Yeah, it's the same issue as `Optional`, `Unsafe*Pointer`, etc

Reply:
Ah right, - argument has to be `Hashable` to be implicitly convertible to `AnyHashable` but generic parameter could also require conformance to another type which only `AnyHashable` conforms to...

Reply:
Looks great! Would it be worth also adding a test case for the conditional conformance example? Admittedly contrived, but another one I thought of was for multiple levels of value-to-optional, e.g:

```
protocol P {}
extension Optional : P where Wrapped == Int? {}

func bar<T : P>(_ x: T) -> T { x }
let i = 0
let _: Int?? = bar(i)
```

Reply:
@hamishknight  Sure I can add that! I'm also going to add overloaded version of it to make sure that we handle cases - when simplification is done during constraint generation and when it happens in the solver.